File: core-algorithms/1_recursion/3_findSubstrings/findSubstrings.py
def findSubstrings(str):
    if(str == ""):
        return [""]
    
    firstChar = str[0]
    restSubstr = findSubstrings(str[1:])
    withChar = list(map(lambda x: firstChar + x, restSubstr))
    return restSubstr + withChar

print(findSubstrings("abc"))

File: core-algorithms/1_recursion/3_findSubstrings/findSubstrings.ts
const findSubstrings = (str: string): Array<string> => {
  if (!str.length) {
    return [""];
  }

  const firstChar = str.charAt(0);
  const restSubstrings = findSubstrings(str.substring(1));
  const withChar = restSubstrings.map((s) => firstChar + s);
  return [...restSubstrings, ...withChar];
};

console.log(findSubstrings("abc"));
